consensus/src/data_manager.rs

```diff
@@ -0,0 +1,235 @@
+                                LogicalTime::new(block.epoch(), block.round()),
+                            )
+                            .await;
+                        proof_with_status
+                            .status
+                            .lock()
+                            .replace(DataStatus::Requested(receivers));
+                    }
+                }
+                Payload::Empty => {}
+                Payload::DirectMempool(_) => {
+                    unreachable!()
+                }
+            }
+        }
+    }
+
+    /// Extract transaction from a given block
+    /// Assumes it is never called for the same block concurrently. Otherwise status can be None.
+    pub async fn get_transactions(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
+        if block.payload().is_none() || block.payload().unwrap().is_empty() {
+            return Ok(Vec::new());
+        }
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            match block.payload().unwrap() {
+                Payload::InQuorumStore(proof_with_status) => {
+                    let status = proof_with_status.status.lock().take();
+                    match status.expect("Should have been updated before") {
+                        DataStatus::Cached(data) => {
+                            proof_with_status
+                                .status
+                                .lock()
+                                .replace(DataStatus::Cached(data.clone()));
+                            Ok(data)
+                        }
+                        DataStatus::Requested(receivers) => {
+                            let mut vec_ret = Vec::new();
+                            debug!("QSE: waiting for data on {} receivers", receivers.len());
+                            for rx in receivers {
+                                match rx.await {
+                                    Err(_) => {
+                                        // We probably advanced epoch already.
+                                        warn!("Oneshot channel to get a batch was dropped");
+                                        let new_receivers = self
+                                            .request_transactions(
+                                                proof_with_status.proofs.clone(),
+                                                LogicalTime::new(block.epoch(), block.round()),
+                                            )
+                                            .await;
+                                        // Could not get all data so requested again
+                                        proof_with_status
+                                            .status
+                                            .lock()
+                                            .replace(DataStatus::Requested(new_receivers));
+                                        return Err(BlockNotFound(block.id()));
```

Comment:
we probably should add new error types for data manager.

Reply:
Should I add them in the same place (enum Error)?

Reply:
Done. We can track the batch id, but it will require some changes.